File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        List<Integer> ans = new ArrayList<>();
        // List<Integer> visited = new ArrayLis<>();
        int m = matrix.length;
        int n = matrix[0].length;
        int top = 0;
        int bottom = m-1;
        int left = 0;
        int right = n-1;

        while(ans.size() < m*n)
        {
            for(int i = left; i<= right; i++)
            {
                ans.add(matrix[top][i]);
            }
            top++;
            for(int i = top; i<= bottom; i++)
            {
                ans.add(matrix[i][right]);
            }
            right--;
            if(top<=bottom)
            {
                for(int i = right; i>=left; i--)
                {
                    ans.add(matrix[bottom][i]);
                }
                bottom--;
            }
            if(left<=right)
            {
                for(int i = bottom; i>=top; i--)
                {
                    ans.add(matrix[i][left]);
                }
                left++;
            }
        }
        return ans;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        m = len(matrix)
        n = len(matrix[0])
        srow = 0
        erow = m
        scol = 0
        ecol = n
        ans = []

        while srow < erow and scol < ecol:
            for j in range(scol, ecol):
                ans.append(matrix[srow][j])
            
            for i in range(srow+1, erow):
                ans.append(matrix[i][ecol-1])
            
            for j in range(ecol - 2, scol-1, -1):
                if srow == erow -1:
                    break
                ans.append(matrix[erow-1][j])

            for i in range(erow-2, srow, -1):
                if scol == ecol - 1:
                    break
                ans.append(matrix[i][scol])

            srow += 1
            erow -= 1
            scol += 1
            ecol -= 1

        return ans